File: deploy/bypass-service/npm/textcli-core/parser.js
// textcli-core parser — Protocol directive parser (SPEC §1.1)
// Zero-dependency, isomorphic with Python textcli-loader parser.py.
//
// Input:  "AI:domain;action,param1,param2"
// Output: {domain, action, params[]} or {error, reason}
//
// Supports:
//   - AI: and 指令: prefixes (equivalent)
//   - JSON objects/arrays in params (bracket-depth tracking)
//   - String quotes inside params (commas inside quotes not split)
//   - Escape sequences in string params
//   - Free-text trailing param (may contain commas)

"use strict";

const _MAX_LENGTH = 2048;
const _MAX_PARAMS = 50;
const _DIRECTIVE_PATTERN = /^\s*(?:AI|指令)[：:]([^;]+);([^,]+)(?:,(.+))?\s*$/;

// ─── Private ─────────────────────────────────────────

/**
 * Split params by comma, respecting JSON brackets and string quotes.
 * Isomorphic with Python textcli-loader _split_params().
 */
function _splitParams(raw) {
  const result = [];
  let buf = "";
  let depth = 0;        // { } and [ ] nesting
  let inString = false;
  let strCh = "";
  let escape = false;

  for (let i = 0; i < raw.length; i++) {
    const ch = raw[i];

    if (escape) {
      buf += ch;
      escape = false;
      continue;
    }

    if (ch === "\\") {
      escape = true;
      continue;
    }

    if ((ch === '"' || ch === "'") && depth === 0) {
      inString = !inString;
      strCh = inString ? ch : "";
      buf += ch;
      continue;
    }

    if (inString) {
      buf += ch;
      continue;
    }

    if (ch === "{" || ch === "[") {
      depth++;
      buf += ch;
      continue;
    }

    if (ch === "}" || ch === "]") {
      depth = Math.max(0, depth - 1);
      buf += ch;
      continue;
    }

    if (ch === "," && depth === 0) {
      const val = buf.trim();
      if (val.length > 0) result.push(val);
      buf = "";
      continue;
    }

    buf += ch;
  }

  const val = buf.trim();
  if (val.length > 0) result.push(val);
  return result;
}

// ─── Public ──────────────────────────────────────────

/**
 * Parse a text-cli directive string.
 *
 * @param {string} prompt - e.g. "AI:weather;query,Beijing"
 * @returns {{domain:string, action:string, params:string[]} | {error:string, reason:string}}
 */
function parse(prompt) {
  if (!prompt || typeof prompt !== "string" || !prompt.trim()) {
    return { error: "INVALID_PARAMS", reason: "prompt is required" };
  }

  prompt = prompt.trim();

  if (prompt.length > _MAX_LENGTH) {
    return { error: "INVALID_PARAMS", reason: `directive exceeds max length (${_MAX_LENGTH})` };
  }

  const match = prompt.match(_DIRECTIVE_PATTERN);
  if (!match) {
    return { error: "INVALID_PARAMS", reason: `invalid directive format: ${prompt}` };
  }

  const domain = match[1].trim();
  const action = match[2].trim();
  const rawParams = match[3] || "";

  const params = rawParams ? _splitParams(rawParams) : [];

  if (params.length > _MAX_PARAMS) {
    return { error: "INVALID_PARAMS", reason: `too many parameters (${params.length}), max ${_MAX_PARAMS}` };
  }

  if (!domain) {
    return { error: "INVALID_PARAMS", reason: "domain is empty" };
  }
  if (!action) {
    return { error: "INVALID_PARAMS", reason: "action is empty" };
  }

  return { domain, action, params };
}

module.exports = { parse };

File: src/skeleton/bypass-service/pypi/src/textcli_loader/parser.py
"""
Directive parser — isomorphic with text-cli service core/parser.py.

Extracted to zero-dependency module for pip-installable textcli-loader.
"""

import re
from dataclasses import dataclass

_DIRECTIVE_PATTERN = re.compile(
    r"^\s*(?:AI|指令)[：:]([^;]+);([^,]+)(?:,(.+))?\s*$"
)
_MAX_LENGTH = 2048
_MAX_PARAMS = 50


@dataclass
class ParsedDirective:
    domain: str
    action: str
    params: list[str]
    raw: str

    @property
    def directive_key(self) -> str:
        """Return directive key preserving original prefix."""
        for prefix in ("AI:", "指令:"):
            if self.raw.startswith(prefix):
                return f"{prefix[:-1]}:{self.domain};{self.action}"
        return f"AI:{self.domain};{self.action}"


class DirectiveParseError(Exception):
    def __init__(self, message: str, code: str = "INVALID_PARAMS"):
        self.message = message
        self.code = code
        super().__init__(message)


def _split_params(raw: str) -> list[str]:
    """Split params by comma, respecting JSON brackets and string quotes."""
    result = []
    buf = []
    depth = 0
    in_string = False
    escape = False
    for ch in raw:
        if escape:
            buf.append(ch)
            escape = False
            continue
        if ch == '\\':
            buf.append(ch)
            escape = True
            continue
        if ch == '"' and depth == 0:
            in_string = not in_string
            buf.append(ch)
            continue
        if in_string:
            buf.append(ch)
            continue
        if ch in ('[', '{'):
            depth += 1
            buf.append(ch)
            continue
        if ch in (']', '}'):
            depth -= 1
            buf.append(ch)
            continue
        if ch == ',' and depth == 0:
            val = ''.join(buf).strip()
            if val:
                result.append(val)
            buf = []
            continue
        buf.append(ch)
    val = ''.join(buf).strip()
    if val:
        result.append(val)
    return result


def parse(prompt: str | None) -> ParsedDirective:
    """Parse a text-cli directive string.

    Raises DirectiveParseError on invalid input.
    """
    if not prompt or not prompt.strip():
        raise DirectiveParseError("prompt is required")

    prompt = prompt.strip()

    if len(prompt) > _MAX_LENGTH:
        raise DirectiveParseError(
            f"directive exceeds max length ({_MAX_LENGTH})"
        )

    match = _DIRECTIVE_PATTERN.match(prompt)
    if not match:
        raise DirectiveParseError(f"invalid directive format: {prompt}")

    domain = match.group(1).strip()
    action = match.group(2).strip()
    raw_params = match.group(3)

    params: list[str] = []
    if raw_params:
        params = _split_params(raw_params)

    if len(params) > _MAX_PARAMS:
        raise DirectiveParseError(
            f"too many parameters ({len(params)}), max {_MAX_PARAMS}"
        )

    if not domain:
        raise DirectiveParseError("domain is empty")
    if not action:
        raise DirectiveParseError("action is empty")

    return ParsedDirective(
        domain=domain,
        action=action,
        params=params,
        raw=prompt,
    )
